### PR TITLE
Buffered process time timers

### PR DESCRIPTION
**Pull request by dxichen**
Addresses #22975

That the added code includes a new KeyTimerData format to be saved in order to facilitate migration to this new buffered format (fully backwards compatible as it still maintains the previous state format for process timers)

Furthermore, with this change PROCESS_TIME timers no longer directly uses the TimerRegistry, rather it passes the ready timers on watermark similar to the EVENT_TIME timers

R: @xinyuiscool 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by dxichen** [comment]
R: @xinyuiscool

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java [hunk 1]
```diff
@@ -307,7 +327,11 @@ public void setTimer(TimerData timerData) {
           break;
 
         case PROCESSING_TIME:
-          timerRegistry.schedule(keyedTimerData, timerData.getTimestamp().getMillis());
+          // Append to buffer iff not full
+          if (processTimeBuffer.size() < maxProcessTimerBufferSize) {
+            processTimeBuffer.add(keyedTimerData);
+          }
```
Comment (mynameborat) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:333]: Do we need to do anything in case we have reached the buffer size? just like the event time scenario.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java [hunk 2]
```diff
@@ -545,57 +596,58 @@ private void reloadEventTimeTimers() {
...
+      if (!keyedTimerIter.hasNext()) {
+        // Migrate from timerState to keyedTimerState
+        while (timersIter.hasNext()) {
+          final Map.Entry<TimerKey<K>, Long> entry = timersIter.next();
+          final KeyedTimerData<K> keyedTimerData =
+              TimerKey.toKeyedTimerData(entry.getKey(), entry.getValue(), timeDomain, keyCoder);
+          keyedTimerState.add(keyedTimerData);
+        }
+      }
+      keyedTimerState.closeIterators();
```
Comment (mynameborat) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:648]: Will this code go away after migration? If so, can we add a TODO and attach a JIRA ticket?

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java [hunk 3]
```diff
@@ -191,11 +193,26 @@ public Collection<KeyedTimerData<K>> removeReadyTimers() {
...
       }
     }
+
+    // Flush all timers for process time
+    final Iterator<KeyedTimerData<K>> processBufferIterator = processTimeBuffer.iterator();
+    while (processBufferIterator.hasNext() && readyTimers.size() < maxReadyTimersToProcessOnce) {
+      KeyedTimerData<K> processTimerData = processBufferIterator.next();
+      readyTimers.add(processTimerData);
+      processBufferIterator.remove();
+      state.deletePersisted(processTimerData);
```
Comment (mynameborat) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:203]: Given we persist the deletes for the process timers, are the ready timers persisted themselves in case of failures?

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java [hunk 4]
```diff
@@ -218,6 +235,10 @@ public Instant getOutputWatermark() {
   // for unit test only
   NavigableSet<KeyedTimerData<K>> getEventTimeBuffer() {
     return eventTimeBuffer;
+  } // todo dchen1
```
Comment (mynameborat) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:238]: remove `//todo dchen1`

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java [hunk 5]
```diff
@@ -279,18 +302,15 @@ public void setTimer(TimerData timerData) {
...
         case EVENT_TIME:
           /*
            * To determine if the upcoming KeyedTimerData could be added to the Buffer while
            * guaranteeing the Buffer's timestamps are all <= than those in State Store to preserve
            * timestamp eviction priority:
            *
-           * <p>1) If maxEventTimeInBuffer == long.MAX_VALUE, it indicates that the State is empty,
-           * therefore all the Event times greater or lesser than newTimestamp are in the buffer;
-           *
-           * <p>2) If newTimestamp < maxEventTimeInBuffer, it indicates that there are entries
```
Comment (mynameborat) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:293]: Why is this deleted? Does it no longer apply for event time is it?

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java [hunk 6]
```diff
@@ -444,6 +470,16 @@ private class SamzaTimerState {
...
                           VarLongCoder.of()));
 
+      this.timestampSortedProcessTimeTimerState =
+          (SamzaSetState<KeyedTimerData<K>>)
+              nonKeyedStateInternalsFactory
+                  .stateInternalsForKey(null)
+                  .state(
+                      StateNamespaces.global(),
+                      StateTags.set(
+                          timerStateId + "-pts",
```
Comment (mynameborat) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:480]: looks like we anyways don't have consistent naming scheme and can't change `-ts` to `-ets`. maybe be more descriptive instead of `pts` to help readability and debuggability.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java [hunk 7]
```diff
@@ -545,57 +596,58 @@ private void reloadEventTimeTimers() {
...
+
+      while (iter.hasNext() && processTimeBuffer.size() < maxProcessTimerBufferSize) {
+        final KeyedTimerData keyedTimerData = iter.next();
+        processTimeBuffer.add(keyedTimerData);
       }
-      processingTimeTimerState.closeIterators();
 
-      LOG.info("Loaded {} processing time timers in memory", count);
+      timestampSortedProcessTimeTimerState.closeIterators();
+      LOG.info("Loaded {} processing time timers in memory", processTimeBuffer.size());
```
Comment (mynameborat) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:609]: Need to close the iterator since the loop can exit early

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java (hunk 1 above)
Comment (dxichen) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:333]: No we don't need to replace any events in the buffer like in event time because process time is fifo only, if the buffer is full we don't do anything and it is simply stored in the state

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java (hunk 2 above)
Comment (dxichen) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:648]: This could be removed after all the samza-beam jobs have migrated to use only keyed timer data, will add a todo

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java (hunk 5 above)
Comment (dxichen) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:293]: This no longer represent the below code completely changed (even before my change, removed 1) and kept 2)

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java (hunk 6 above)
Comment (dxichen) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:480]: good point will name it to process-timestamp-sorted

**Comment by dxichen** [comment]
Run Java PreCommit

File: runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaPipelineOptions.java [hunk 8]
```diff
@@ -80,7 +80,13 @@ public interface SamzaPipelineOptions extends PipelineOptions {
   @Default.Integer(50000)
   int getEventTimerBufferSize();
 
-  void setEventTimerBufferSize(int eventTimerBufferSize);
+  void setEventTimerBufferSize(int processingTimerBufferSize);
```
Comment (kw2542) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaPipelineOptions.java:83]: argument should be named `eventTimerBufferSize`

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java (hunk 3 above)
Comment (dxichen) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java]: due to samza checkpointing with changelog/blobstore, and state on restart from failure is read from the checkpoint any persisted deletes will be rolled back

**Comment by dxichen** [comment]
Run Samza ValidatesRunner

**Comment by kw2542** [comment]
Run Java Samza PortableValidatesRunner

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java [hunk 9]
```diff
@@ -193,7 +201,7 @@ public Collection<KeyedTimerData<K>> removeReadyTimers() {
     }
     LOG.debug("Removed {} ready timers", readyTimers.size());
 
-    if (readyTimers.size() == maxReadyTimersToProcessOnce
+    if (readyTimers.size() >= maxReadyTimersToProcessOnce
```
Comment (kw2542) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:204]: This change may not be necessary since readyTimers size is incremented by one per loop.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java [hunk 10]
```diff
@@ -279,35 +298,49 @@ public void setTimer(TimerData timerData) {
...
+          if ((maxEventTimerBufferSize > eventTimeBuffer.size() && !eventsTimersInState)
+              || newTimestamp < eventTimeBuffer.last().getTimerData().getTimestamp().getMillis()) {
             eventTimeBuffer.add(keyedTimerData);
             if (eventTimeBuffer.size() > maxEventTimerBufferSize) {
               eventTimeBuffer.pollLast();
-              maxEventTimeInBuffer =
-                  eventTimeBuffer.last().getTimerData().getTimestamp().getMillis();
             }
+          } else {
+            eventsTimersInState = true;
```
Comment (xinyuiscool) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java:323]: The timers are always in state. The flag name is just wrong.

**Review by xinyuiscool — COMMENTED** [review]
I feel the changes to use boolean flags to track whether there are extra times in the state is much harder to understand than the previous maxTimeInBuffer. The maxTimerInBuffer is quite simple to read and the condition is also simpler to understand. I would not suggest making such a change without any benefit other than making the logic 10X hard to understand.
